File: Leetcode/TopInterview150/242.ValidAnagram/ValidAnagram.py
'''
Given two strings s and t, return true if t is an anagram of s, and false otherwise.

An Anagram is a word or phrase formed by rearranging the letters of a different word 
or phrase, typically using all the original letters exactly once.

 

Example 1:

Input: s = "anagram", t = "nagaram"
Output: true
Example 2:

Input: s = "rat", t = "car"
Output: false


Constraints:

1 <= s.length, t.length <= 5 * 104
s and t consist of lowercase English letters.
'''

class Solution(object):
    def isAnagram(self, s, t):
        """
        :type s: str
        :type t: str
        :rtype: bool
        """
        #first lower the strings in case thery not
        s = s.lower()
        t = t.lower()

        #build the hash maps
        dict1 = {}
        dict2 = {}

        #First we check if we have the same amount of
        #letters in each word
        if len(dict1) != len(dict2):
            return False

        #fill the dict1 with s
        for i in s:
            if i in dict1:
                dict1[i] += 1
            else:
                dict1[i] = 1
        
        #fill the dict2 with t
        for j in t:
            if j in dict2:
                dict2[j] += 1
            else:
                dict2[j] = 1

        if dict1 == dict2:
            return True
        
        return False

File: Leetcode/TopInterview150/242.ValidAnagram/ValidAnagram.java
/*

Given two strings s and t, return true if t is an anagram of s, and false otherwise.

 

Example 1:

Input: s = "anagram", t = "nagaram"

Output: true

Example 2:

Input: s = "rat", t = "car"

Output: false

 

Constraints:

1 <= s.length, t.length <= 5 * 104
s and t consist of lowercase English letters.

*/


class Solution {
    public boolean isAnagram(String s, String t) {
        if (t.length() != s.length()) return false;

        // Lower the strings
        s.toLowerCase();
        t.toLowerCase();

        // Delcare to empty hashmaps
        Map<Character,Integer> map1 = new HashMap ();
        Map<Character,Integer> map2 = new HashMap ();

        // Fill first and second hashmap with s and t {character:count}
        // at the same time for not repeating code
        for (int i = 0; i < s.length(); i++) {
            // Get ith character of string
            char cs = s.charAt(i);
            char ct = t.charAt(i);

            // put in the value the current value plus one
            // but if it doesn't exist add it and add one
            map1.put(cs,map1.getOrDefault(cs,0)+1);
            map2.put(ct,map2.getOrDefault(ct,0)+1);
        }

        // Compare two hashmaps (key and value)
        if (map1.equals(map2)) return true;

        return false;
    }
}

// Time complexity O(n)
